src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1008,11 +1008,29 @@ int32_t SystemNative_PosixFAdvise(intptr_t fd, int64_t offset, int64_t length, i
 #endif
 }
 
-int32_t SystemNative_PosixFAllocate(intptr_t fd, int64_t offset, int64_t length)
+int32_t SystemNative_PosixFAllocate(intptr_t fd, int64_t offset, int64_t length, int32_t mode)
 {
     assert_msg(offset == 0, "Invalid offset value", (int)offset);
 
     int fileDescriptor = ToFileDescriptor(fd);
+#if !HAVE_POSIX_FALLOCATE64 && !HAVE_POSIX_FALLOCATE
+    // posix_fallocate does not allow for shrinking file
+    // when it's not available, we need to ensure that the file won't be shrinked
```

Comment:
Did you consider having this in the managed side? Seems to me that may be clearer to have it in `FileStreamHelpers.ShouldPreallocate`.

Reply:
In managed layer we don't have information whether `posix_fallocate` is implemented or not. In theory I could use `!OperatingSystem.IsLinux()` condition, but it would be only a guess.

Reply:
```suggestion
    // posix_fallocate does not allow for shrinking files.
    // When it's not available, we need to ensure that the file won't be shrunk.
```